packages/hardhat-viem/src/internal/bytecode.ts

```diff
@@ -0,0 +1,145 @@
+        )
+      );
+    } else if (matchingLibraries.length === 0) {
+      throw new UnnecessaryLibraryLinkError(contractName, linkedLibraryName);
+    }
+  }
+}
+
+export async function throwOnMissingLibrariesAddress(
+  contractName: string,
+  libraries: Libraries<viemT.Address>,
+  neededLibraries: Array<Pick<Link, "libraryName" | "sourceName">>
+) {
+  const missingLibraries = [];
+  for (const { sourceName, libraryName } of neededLibraries) {
+    const address =
+      libraries[`${sourceName}:${libraryName}`] ?? libraries[libraryName];
+
+    if (address === undefined) {
+      missingLibraries.push({ sourceName, libraryName });
+    }
+  }
+
+  if (missingLibraries.length > 0) {
+    throw new MissingLibraryAddressError(contractName, missingLibraries);
+  }
+}
+
+export async function throwOnInvalidLibrariesAddress(
+  contractName: string,
+  libraries: Libraries<viemT.Address>,
+  neededLibraries: Link[]
+) {
+  const librariesWithInvalidAddress = [];
+  for (const { sourceName, libraryName } of neededLibraries) {
+    const address =
+      libraries[`${sourceName}:${libraryName}`] ?? libraries[libraryName];
+
+    if (address === undefined) {
+      librariesWithInvalidAddress.push({ sourceName, libraryName, address });
+    }
+  }
+
+  if (librariesWithInvalidAddress.length > 0) {
+    throw new InvalidLibraryAddressError(
+      contractName,
+      librariesWithInvalidAddress
+    );
+  }
+}
+
+export async function throwOnOverlappingLibraryNames(
+  contractName: string,
+  libraries: Libraries<viemT.Address>,
+  neededLibraries: Array<Pick<Link, "libraryName" | "sourceName">>
+) {
+  for (const { sourceName, libraryName } of neededLibraries) {
+    if (
+      libraries[`${sourceName}:${libraryName}`] !== undefined &&
+      libraries[libraryName] !== undefined
+    ) {
+      throw new OverlappingLibraryNamesError(sourceName, libraryName);
+    }
+  }
+}
+
+export async function resolveBytecodeWithLinkedLibraries(
+  artifact: Artifact,
+  libraries: Libraries<viemT.Address>
+): Promise<string> {
+  const { linkReferences } = artifact;
+
+  const neededLibraries: Array<Omit<Link, "address">> = [];
+  for (const [sourceName, sourceLibraries] of Object.entries(linkReferences)) {
+    for (const [libraryName] of Object.entries(sourceLibraries)) {
+      neededLibraries.push({
+        sourceName,
+        libraryName,
+      });
+    }
+  }
+
+  await throwOnAmbigousLibraryNameOrUnnecessaryLink(
+    artifact.contractName,
+    libraries,
+    neededLibraries
+  );
+  await throwOnOverlappingLibraryNames(
+    artifact.contractName,
+    libraries,
+    neededLibraries
+  );
+  await throwOnMissingLibrariesAddress(
+    artifact.contractName,
+    libraries,
+    neededLibraries
+  );
+
+  const resolvedLibraries: Link[] = neededLibraries.map((link) => ({
+    ...link,
+    address:
+      libraries[`${link.sourceName}:${link.libraryName}`] ??
+      libraries[link.libraryName],
+  }));
```

Comment:
I think resolving the address when you're pushing the objects to `neededLibraries` in the initial `for..of` loop would make the code cleaner. That way we can use the type `Link` everywhere and get rid of `throwOnInvalidLibrariesAddress` which is basically a duplicate of `throwOnMissingLibrariesAddress`.